MIPs/mip-2.md

```diff
@@ -0,0 +1,117 @@
+---
+MIP: 1
+Title: Implement `wallet_swap` to support Dapps to swap
+Status: Draft
+Stability: Experimental
+discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions
+Author(s): Tomás Santos (@tommasini)
+Type: Maintainer
+Created: 2023-11-10
+---
+
+## Summary
+
+This proposal introduces a new JSON RPC method `wallet_swap`` that allows dApps to request MetaMask to perform a token swap operation. The method accepts parameters for the source and destination tokens and initiates a token swap operation.
+
+## Motivation
+
+The motivation behind this MIP is to provide a standardized way for dApps to request token swaps from MetaMask, improving interoperability and user experience.
+
+# Usage Example
+
+```
+ethereum.request({
+  method: 'wallet_swap',
+  params: {
+    from: {
```

Comment:
I think this should include the `from` `address` like `sendTransaction` does, since you can `sendTransaction` with any account you have connected to the site, the same logic should be applied here.

Reply:
Amazing suggestion! Answered on this [comment](https://github.com/MetaMask/metamask-improvement-proposals/pull/25#issuecomment-1827794826) below:

We can use the same logic as we use on the send transaction rpc method, and use the account connected to the dapp, adding the from field that way!

In the other hand it would be nice to use the current screens with the swaps logic that already exist, what do you think if we ask the user to switch the selected account to account to the new from address field (if it's different than the one selected) instead of creating a new screen with all the swaps logic? (We can also do the first implementation since it's quicker and re-uses logic, and improve it to the second one if we see a big adoption, in order to improve it)

Curious, what do you think it's the best name for the property, from_address, user_address or just address?

Reply:
you can `eth_requestAccounts` or `requestPermissions` and select multiple accounts that get connected to the dapp. that means `sendTransaction` as well as any other account based action needs to take the account as a param.

as for naming, just looking around the space for some naming ideas I found:

`in` and `out`

https://docs.uniswap.org/sdk/v3/guides/swaps/quoting


`buyToken`, `sellToken` and they use `allowanceTarget` as the address
https://0x.org/docs/0x-swap-api/guides/swap-tokens-with-0x-swap-api#example-request


after seeing these I think I'm leaning towards `from`, `to` and `address` or maybe `user_address`.

Reply:
Thanks! Yes indeed! I will do a refactor and include this as well! I understood now that you were not talking about the flow on the product itself but on the `sendTransaction` rpc method right?